### PR TITLE
Switch to Pint Units

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description
The aim of this PR is to switch over to using `pint` units. The goal is to remove a number of awkward behaviors, including

  - a `dimensionless` quantity losing its unit after scalar multiplication.
  - handle cases such as `(x * simtk.unit.joule) / (y * simtk.unit.calorie)` being dimensionless.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] Add conversions between `simtk.unit.X` <-> `pint.X`
  - [x] Add lots of tests for the conversions.
  - [x] Replace most instances of simtk.unit
  - [x] Run regression tests.

## Status
- [x] Ready to go

### CODE REVIEW

File: propertyestimator/__init__.py [hunk 1]
```diff
@@ -24,6 +25,10 @@
         formatted_exception = traceback.format_exception(None, e, e.__traceback__)
         logging.warning(f'Could not load the {entry_point} plugin: {formatted_exception}')
 
+# Set up pint.
+from pint import UnitRegistry
+unit = UnitRegistry()
```
Comment (j-wags) [inline propertyestimator/__init__.py]: This is very clever, but I worry that it'll be really confusing for a contributor some day.

File: propertyestimator/backends/backends.py [hunk 2]
```diff
@@ -108,7 +109,7 @@ def wallclock_time_limit(self):
         return self._wallclock_time_limit
 
     def __init__(self, number_of_threads=1, number_of_gpus=0, preferred_gpu_toolkit=None,
-                 per_thread_memory_limit=1*(unit.giga*unit.bytes), wallclock_time_limit="01:00"):
+                 per_thread_memory_limit=1*unit.gigabytes, wallclock_time_limit="01:00"):
```
Comment (j-wags) [inline propertyestimator/backends/backends.py]: Now _that_ is really cool.

File: propertyestimator/backends/backends.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 import re
 from enum import Enum
 
-from simtk import unit
+from propertyestimator import unit as unit
```
Comment (j-wags) [inline propertyestimator/backends/backends.py]: In most of the other files, this doesn't have the `as unit` part at the end. Does that make a difference?

File: propertyestimator/backends/dask.py [hunk 4]
```diff
@@ -385,7 +384,7 @@ def start(self):
         from dask_jobqueue.lsf import lsf_detect_units, lsf_format_bytes_ceil
 
         requested_memory = self._resources_per_worker.per_thread_memory_limit
-        memory_bytes = requested_memory.value_in_unit(unit.byte)
+        memory_bytes = requested_memory.to(unit.byte).magnitude
```
Comment (j-wags) [inline propertyestimator/backends/dask.py]: I like pint already.

File: propertyestimator/protocols/simulation.py [hunk 5]
```diff
@@ -10,12 +10,14 @@
...
 import yaml
-from simtk import unit, openmm
+from simtk import openmm, unit as simtk_unit
 from simtk.openmm import app
 
+from propertyestimator import unit
 from propertyestimator.thermodynamics import ThermodynamicState, Ensemble
 from propertyestimator.utils.exceptions import PropertyEstimatorException
-from propertyestimator.utils.openmm import setup_platform_with_resources
+from propertyestimator.utils.openmm import setup_platform_with_resources, openmm_quantity_to_pint,\
```
Comment (j-wags) [inline propertyestimator/protocols/simulation.py]: I think it's becoming preferred to do multi-line imports using parentheses: https://www.python.org/dev/peps/pep-0328/

File: propertyestimator/protocols/simulation.py [hunk 6]
```diff
@@ -394,33 +401,36 @@ def _write_statistics_array(self, raw_statistics, temperature, pressure,
...
         # systems degrees of freedom.
-        temperatures = 2.0 * kinetic_energies / (degrees_of_freedom * unit.MOLAR_GAS_CONSTANT_R)
+        temperatures = 2.0 * kinetic_energies / (degrees_of_freedom * unit.molar_gas_constant)
 
         # Calculate the systems enthalpy and reduced potential.
         total_energies = potential_energies + kinetic_energies
         enthalpies = None
 
-        reduced_potentials = potential_energies / unit.AVOGADRO_CONSTANT_NA
+        reduced_potentials = potential_energies / unit.avogadro_number
```
Comment (j-wags) [inline propertyestimator/protocols/simulation.py]: (Not a suggestion for this PR, but maybe a discussion we should have about standardizing design)
I'm worried that we're going to reach a reckoning some day with the fact that we use `kJ/mole` to mean two completely separate things:
1) A very small amount of energy, like "a microfemtojoule (or whatever the proper prefix is)", that's useful for describing amounts of energy like those that a molecule encounters. 
2) An actual unit of energy divided by a unit of quantity, like "this dynamite has 100 kJ/mol of energy. If you add more moles, you get more energy."

File: propertyestimator/tests/test_utils/test_openmm.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    openmm_raw_value = openmm_quantity.value_in_unit(simtk_unit.gram / simtk_unit.mole)
+
+    pint_quantity = openmm_quantity_to_pint(openmm_quantity)
+    pint_raw_value = pint_quantity.to(unit.gram / unit.mole).magnitude
+
+    assert np.allclose(openmm_raw_value, pint_raw_value)
+
+
+@pytest.mark.parametrize("openmm_unit", _get_all_openmm_units())
+@pytest.mark.parametrize("value", [random(), randint(1, 10), [random(), random()], np.array([random(), random()])])
```
Comment (j-wags) [inline propertyestimator/tests/test_utils/test_openmm.py]: This is interesting. If this is going to fail for particular values of `value`, does it make sense that we should see that failure during testing at a random time in the future (probably during unrelated development)? I think it would be better to have a --runslow option that tests a pretty exhaustive set of values, but during normal testing just have this regularly run with a fixed value. [1]

[1] The fixed value could still be random :-P https://www.xkcd.com/221/

File: propertyestimator/tests/test_utils/test_openmm.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    pint_quantity = openmm_quantity_to_pint(openmm_quantity)
+    pint_raw_value = pint_quantity.magnitude
+
+    assert np.allclose(openmm_raw_value, pint_raw_value)
+
+
+@pytest.mark.parametrize("pint_unit", _get_all_pint_units())
+@pytest.mark.parametrize("value", [random(), randint(1, 10), [random(), random()], np.array([random(), random()])])
+def test_pint_to_openmm(pint_unit, value):
```
Comment (j-wags) [inline propertyestimator/tests/test_utils/test_openmm.py]: These are really cool tests.

File: propertyestimator/tests/test_utils/test_openmm.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+            pint_quantity = openmm_quantity_to_pint(openmm_quantity)
+            pint_raw_value = pint_quantity.magnitude
+
+            assert np.isclose(openmm_raw_value, pint_raw_value)
+
+
+@pytest.mark.parametrize("openmm_unit", {*_get_all_openmm_units(), simtk_unit.dalton**2, simtk_unit.dalton**3})
+@pytest.mark.parametrize("value", [random(), randint(1, 10), [random(), random()], np.array([random(), random()])])
+def test_openmm_unit_conversions(openmm_unit, value):
```
Comment (j-wags) [inline propertyestimator/tests/test_utils/test_openmm.py]: 👍

File: propertyestimator/tests/test_utils/test_openmm.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                                            isinstance(openmm_quantity, list) or
+                                            isinstance(openmm_quantity, np.ndarray)):
+        openmm_raw_value = openmm_quantity
+    else:
+        openmm_raw_value = openmm_quantity.value_in_unit(openmm_unit)
+
+    assert np.allclose(openmm_raw_value, pint_raw_value)
+
+
+def test_constants():
```
Comment (j-wags) [inline propertyestimator/tests/test_utils/test_openmm.py]: I _really_ like how well thought-out these tests are.

File: propertyestimator/__init__.py (hunk 1 above)
Comment (SimonBoothroyd) [reply propertyestimator/__init__.py]: Unfortunately this is just the pint way of doing things: https://pint.readthedocs.io/en/0.9/tutorial.html#using-pint-in-your-projects

File: propertyestimator/backends/backends.py (hunk 3 above)
Comment (SimonBoothroyd) [reply propertyestimator/backends/backends.py]: Typo - thanks!

File: propertyestimator/protocols/simulation.py (hunk 5 above)
Comment (SimonBoothroyd) [reply propertyestimator/protocols/simulation.py]: Agree - will keep this for now as it's consistent, and open a PR to fix this across the codebase in the future

File: propertyestimator/tests/test_utils/test_openmm.py (hunk 7 above)
Comment (SimonBoothroyd) [reply propertyestimator/tests/test_utils/test_openmm.py]: This should _never_ fail in practice. If it does - it means we've found a really weird edge case that would be good to know about and `random` allows us to be on the hunt for this (even though I think it will never find any issues)
